### PR TITLE
`Integrated code lifecycle`: Allow admins to pause build agents

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

We want to give admins the possibility to pause build agents. Pausing a build agent means that the build agent will no longer process any build jobs. If the agent is currently processing build jobs, there will be a grace period where it will wait for the jobs to finish. Afterward, the jobs will be canceled and added back to the queue. The canceled jobs won't produce a result.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Student
- 1 Java exercise
## Scenario 1: Build agent with no running jobs

1. Make sure that code submissions work normally.
2. Go to Build Agents > Click on a build agent > Pause
3. Submit once more and make sure that the pause agent is not processing the job (in multi node setup, you may have to submit a couple of time to make sure that build agent is not processing)
4. Resume Build agent > Submit a few times make sure that the agent is processing jobs.

## Scenario 2: Build agent processing jobs (shorter than grace period)

1. Make a submission > Wait around 5-10 secs and then pause the build agent processing the job (you can find which agent in the Build overview page)
2. Check that the job wasn't canceled and that it produces a result

## Scenario 3: Build agent processing jobs (longer than grace period)
1. Edit a build script of a programming Exercise to make it sleep for 20 secs (add `sleep 20` somewhere in the bash script)
2. Make a submission for that exercise and pause the build agent
3. Make sure that the job is canceled (canceled jobs will be visible in the build agent details view but won't be visible in the build overview page_

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [ ] Test 2

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| BuildAgentInformation.java | 100% | ✅                |
| SharedQueueProcessingService.java | 80% | ✅               |
| BuildJobManagementService.java | 66% | ✅               |
| AdminBuildJobQueueResource.java | 100% | ✅               |
| SharedQueueManagementService.java | 93% | ✅               |
| AdminBuildJobQueueResource.java | 100% | ✅               |
| build-agent.model.ts | 100% | ✅ |
| build-agent-details.component.ts | 96.61% | ✅ ❌ |
| build-agents.service.ts | 100% | ✅ |

### Screenshots

Pausing the agent
image

Resuming the agent
image

Paused status
image

## Release Notes

- **New Features**
  - Introduced a pause and resume mechanism for build agents, allowing for better control during operations.
  - Added new HTTP endpoints for pausing and resuming build agents via the admin API.

- **Bug Fixes**
  - Improved handling of job processing during pause/resume actions.

- **Tests**
  - Enhanced test coverage for the new pause and resume functionalities, ensuring robust performance.
  - Added integration tests for build agent management.
  - Introduced tests to verify build job priorities under various conditions.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:419]: Give me a code suggestion

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:435]: Give me a code suggestion

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:397]:
```suggestion
        if (isPaused) {
```
You don't use `this` elsewhere in the code

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:417]: This is not always 10 seconds, is it?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:419]: Can we also wait for the build job with a timeout of `pauseGracePeriodSeconds * 1000L`? This way we do not have to wait the full time in case the build job finishes earlier.

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:99]: What is the difference between these two attributes?

File: src/main/webapp/i18n/de/buildAgents.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/buildAgents.json:20]: Aren't the 10 seconds configurable?

File: src/main/webapp/i18n/de/buildAgents.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/buildAgents.json:23]: Where are these two translations used?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
image
Nice feature :+1: Probably it makes sense to only resume/pause when it's either in idle or paused state. So don't allow resume and pause at the same time

**Review by coolchock — DISMISSED** [review]
tested scenario 3 on ts3 in the testing session, works as expected

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:99]: processResults is set after the grace period is over. This is necessary to make sure that build jobs that ended during the build jobs produce a result

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:3]:
```suggestion
        <div class="mb-4">
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:4]:
```suggestion
            <div class="d-flex align-items-center justify-content-between mb-3">
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:6]:
```suggestion
                    <h3 class="mb-0" id="build-agent-heading" jhiTranslate="artemisApp.buildAgents.details"></h3>
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:8]:
```suggestion
                    <h3 id="build-agent-name" class="mb-0 ms-3">{{ buildAgent.name }}</h3>
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:10]:
```suggestion
                <div class="ms-3 d-flex justify-content-center gap-2">
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:14]:
```suggestion
                            <span class="d-none d-md-inline ms-1" jhiTranslate="artemisApp.buildAgents.resume"></span>
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:19]:
```suggestion
                            <span class="d-none d-md-inline ms-1" jhiTranslate="artemisApp.buildAgents.pause"></span>
```

**Review by edkaya — CHANGES_REQUESTED** [review]
Overall looks good to me. Just left some code suggestions to handle heading styling in a better way for smaller screens (Current heading style is a little bit weird when the screen is small, see screenshot). Instead of creating random padding/margin values, I think a better approach would be to use bootstrap classes here. I also suggested that we can only show the icons for the buttons if the screen is small.

Screenshot 2024-09-24 at 19 23 42

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [reply src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:6]: mb-0 needed, otherwise align-items-center wont work due to the additional margin of h3

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [reply src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:8]: mb-0 needed, otherwise align-items-center wont work due to the additional margin of h3

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [reply src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:10]: To display buttons next to each other even the screen is small

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [reply src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:14]: Hide labels of buttons, only show icons for smaller screens

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html
Comment (edkaya) [reply src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.html:19]: Hide labels of buttons, only show icons for smaller screens

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.ts
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.ts:32]: I would suggest we make all icons here readonly
```suggestion
    readonly faPause = faPause;
```

File: src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.ts
Comment (edkaya) [inline src/main/webapp/app/localci/build-agents/build-agent-details/build-agent-details/build-agent-details.component.ts:137]: Same here
```suggestion
    resumeBuildAgent(): void {
```

**Review by edkaya — DISMISSED** [review]
Thanks for addressing the feedback. Code lgtm now

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Nice addition. I like that there only is one button now, either resume or pause. I think it would also make sense to disable the button, when a request was sent, and the build agent is not yet in the new state.
image
Triggering several pause/resumes doesn't make sense in my opinion.

**Review by SimonEntholzer — DISMISSED** [review]

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code after merge

**Review by coolchock — APPROVED** [review]
re-approve after merge conflict

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by JohannesStoehr — APPROVED** [review]
New tests also look good to me

**Review by edkaya — APPROVED** [review]
Reapprove after merge conflict

**Review by edkaya — APPROVED** [review]

**Review by edkaya — DISMISSED** [review]
Reapprove after merge confict

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by SimonEntholzer — APPROVED** [review]
Tested again on Ts3, reapprove

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by edkaya — DISMISSED** [review]
Looks good to me, reapprove

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by SimonEntholzer — APPROVED** [review]
reapprove after tests fix

**Pull request merged**
